Http cache partitioning (#9038)
google/WebFundamentals: src/content/en/updates/2020/10/http-cache-partitioning.md

project_path: /web/_project.yaml
book_path: /web/updates/_book.yaml
description: Chrome's http cache partitioning helps with better security and privacy.

> {# wf_published_on: 2020-10-01 #}
> {# wf_updated_on: 2020-10-01 #}

Reviewer: ```suggestion
{# wf_published_on: 2020-10-06 #}
{# wf_updated_on: 2020-10-06 #}
```

> description: Chrome's http cache partitioning helps with better security and privacy.
> 
> {# wf_published_on: 2020-10-06 #}
> {# wf_updated_on: 2020-10-06 #}
> {# wf_featured_image: /web/updates/images/generic/sd-card.png #}
> {# wf_tags: storage-isolation, cache #}
> {# wf_featured_snippet: Chrome's http cache partitioning helps with better security and privacy. #}

Reviewer 2: ```suggestion
description: Chrome's HTTP cache partitioning helps with better security and privacy.

{# wf_published_on: 2020-10-06 #}
{# wf_updated_on: 2020-10-06 #}
{# wf_featured_image: /web/updates/images/generic/sd-card.png #}
{# wf_tags: storage-isolation, cache #}
{# wf_featured_snippet: Chrome's HTTP cache partitioning helps with better security and privacy. #}
```

{# wf_blink_components: Blink>Network #}

# Gaining security and privacy by partitioning the cache {: .page-title }

{% include "web/_shared/contributors/agektmr.html" %}

In general, caching can improve performance by storing data so future requests
for the same data are served faster. For example, a cached resource from the
network can avoid a round trip to the server. A cached computational result can
omit the time to do the same calculation.

In Chrome, the cache mechanism is used in various ways and HTTP Cache is one example.

## How Chrome's HTTP Cache currently works

As of version 85, Chrome caches resources fetched from the network, using their
respective resource URLs as the cache key. (A cache key is used to identify a

> cached resource.)

Reviewer: ```suggestion
cached resource.) A cache key is  a tuple of URLs rather than a single URL.
```

Reviewer: This goes with the next request to make it clear what the key is and how it works.

Author: OK, I see what you mean. Please have a look at how this will look on the website: https://pr-9038-dot-web-central.appspot.com/web/updates/2020/10/http-cache-partitioning
I'm consistent on the tuples below the image while `+` on the paragraph to the right.

Author: At this point we are only talking about status quo which is not using tuples yet. If we were to add this phrase, somewhere after "How will cache partitioning affect Chrome's HTTP Cache?".

Reviewer: Yes.

Reviewer: Yes about where to put the tuple explanation.

that the browser has accessed the same resource in the past, which opens the
browser to security and privacy attacks, like the following:

- **Detect if a user has visited a specific site**: An adversary can detect a
  user's browsing history by checking if the cache has a resource which might be
  specific to a particular site or cohort of sites.
- **[Cross-site search
  attack](https://portswigger.net/daily-swig/new-xs-leak-techniques-reveal-fresh-ways-to-expose-user-information)**:
  An adversary can detect if an arbitrary string is in the user's search results
  by checking whether a 'no search results' image used by a particular website
  is in the browser's cache.
- **Cross-site tracking**: The cache can be used to store cookie-like
  identifiers as a cross-site tracking mechanism.

To mitigate these risks, Chrome will partition its HTTP cache starting in Chrome 86.

## How will cache partitioning affect Chrome's HTTP Cache?

With cache partitioning, cached resources will be keyed using a new "Network
Isolation Key" in addition to the resource URL. The Network Isolation Key is
composed of the top-level site and the current-frame site.

> Note: The "site" is recognized using "scheme://eTLD+1" so if requests are from

Reviewer: ```suggestion
Note: The "site" is recognized using "[scheme://eTLD+1](https://web.dev/same-site-same-origin/)" so if requests are from
```

different pages, but they have the same scheme and effective top-level domain+1
they will use the same cache partition. To learn more about this, read
[Understanding "same-site" and
"same-origin"](https://web.dev/same-site-same-origin/).

Look again at the previous example to see how cache partitioning works in
different contexts:

<figure class="attempt-left">
  <img src="/web/updates/images/2020/10/http-cache-partitioning-1.png">
  <figcaption>
    <b>Cache Key</b>: { <code>https://a.example</code>, <code>https://a.example</code>, <code>https://x.example/doge.png</code> }
  </figcaption>
</figure>

A user visits a page (`https://a.example`) which requests an image
(`https://x.example/doge.png`). In this case, the image is requested from the

> network and cached using `https://a.example` (the top-level site) +
> `https://a.example` (the current-frame site) + `https://x.example/doge.png` (the

Reviewer: ```suggestion
network and cached using a tuple consisting of `https://a.example` (the top-level site),
`https://a.example` (the current-frame site), and `https://x.example/doge.png` (the
```

Reviewer: If you can't use a tuple literal in the text, then use text conventions.

resource URL) as the key. (Note that when the resource request is from the
top-level -frame, the top-level site and current-frame site in the Network
Isolation Key are the same.)

<div class="clearfix"></div>

<figure class="attempt-left">
  <img src="/web/updates/images/2020/10/http-cache-partitioning-2.png">
  <figcaption>
    <b>Cache Key</b>: { <code>https://b.example</code>, <code>https://b.example</code>, <code>https://x.example/doge.png</code> }
  </figcaption>
</figure>

The same user visits a different page (`https://b.example`) which requests the
same image (`https://x.example/doge.png`). Though the same image was loaded in
the previous example, since the key doesn't match it will not be a cache hit.

> The image is requested from the network and cached using `https://b.example` +
> `https://b.example` + `https://x.example/doge.png` as the key.

Reviewer: ```suggestion
The image is requested from the network and cached using a tuple consisting of `https://b.example`,
`https://b.example`, and `https://x.example/doge.png` as the key.
```

<div class="clearfix"></div>

<figure class="attempt-left">
  <img src="/web/updates/images/2020/10/http-cache-partitioning-6.png">
  <figcaption>
    <b>Cache Key</b>: { <code>https://a.example</code>, <code>https://a.example</code>, <code>https://x.example/doge.png</code> }
  </figcaption>
</figure>

Now the user comes back to `https://a.example` but this time the image
(`https://x.example/doge.png`) is embedded in an iframe. In this case, the

> key is `https://a.example` + `https://a.example` + `https://x.example/doge.png`

Reviewer: ```suggestion
key is a tuple containing `https://a.example`, `https://a.example`, and `https://x.example/doge.png`
```

and a cache hit occurs. (Note that when the top-level site and the iframe are
the same site, the resource cached with the top-level frame can be used.

<div class="clearfix"></div>

<figure class="attempt-left">
  <img src="/web/updates/images/2020/10/http-cache-partitioning-4.png">
  <figcaption>
    <b>Cache Key</b>: { <code>https://a.example</code>, <code>https://c.example</code>, <code>https://x.example/doge.png</code> }
  </figcaption>
</figure>

The user is back at `https://a.example` but this time the image is hosted in an
iframe from `https://c.example`.

In this case, the image is downloaded from the network because there is no

> resource in the cache that matches the key `https://a.example` +
> `https://c.example` + `https://x.example/doge.png`.

Reviewer: ```suggestion
resource in the cache that matches the key consisting of `https://a.example`,
`https://c.example`, and `https://x.example/doge.png`.
```

project_path: /web/_project.yaml
book_path: /web/updates/_book.yaml
description: Chrome's http cache partitioning helps with better security and privacy.

{# wf_published_on: 2020-10-01 #}
{# wf_updated_on: 2020-10-01 #}
{# wf_featured_image: /web/updates/images/generic/sd-card.png #}
{# wf_tags: storage-isolation, cache #}
{# wf_featured_snippet: Chrome's http cache partitioning helps with better security and privacy. #}
{# wf_blink_components: Blink>Network #}

# Gaining security and privacy by partitioning the cache {: .page-title }

{% include "web/_shared/contributors/agektmr.html" %}

In general, caching can improve performance by storing data so future requests
for the same data are served faster. For example, a cached resource from the
network can avoid a round trip to the server. A cached computational result can
omit the time to do the same calculation.

> In Chrome, the cache mechanism is used in various ways and HTTP Cache is one of
> such examples.

Reviewer: ```suggestion
In Chrome, the cache mechanism is used in various ways and HTTP Cache is one example.
```

Reviewer: What you wrote was grammatically correct. It just wasn't how we would usually say it.

## How Chrome's HTTP Cache currently works

As of version 85, Chrome caches resources fetched from the network, using their
respective resource URLs as the cache key. (A cache key is used to identify a
cached resource.)

The following example illustrates how a single image is cached and treated in
three different contexts:

<figure class="attempt-left">
  <img src="/web/updates/images/2020/10/http-cache-partitioning-1.png">
  <figcaption>

>     <b>Cache Key</b>: { <code>https://x.example/doge.png</code> }

Reviewer: What are the braces ({}) for?

Author: This expresses a tuple of cache keys.

Reviewer: Can you please put the braces inside the `<code>` elements.

Author: No. I think it’s consistent in style and it looks ok. Also, if I do that for triple keys, the line overflows or breaks.

Reviewer: Most of them do that already. If you don't put them inside the `<code>` element it's not obvious that you're showing a tuple.

image (`https://x.example/doge.png`).  
The browser checks its HTTP Cache to see
if it already has this resource cached, using the image URL as the key. The
browser finds a match in its Cache, so it uses the cached version of the
resource.

<div class="clearfix"></div>

<figure class="attempt-left">
  <img src="/web/updates/images/2020/10/http-cache-partitioning-3.png">
  <figcaption>
    <b>Cache Key</b>: { <code>https://x.example/doge.png</code> }
  </figcaption>
</figure>

It doesn't matter if the image is loaded from within an iframe. If the user
visits another website (`https://c.example`) with an iframe
(`https://d.example`) and the iframe requests the same image
(`https://x.example/doge.png`), the browser can still load the image from its
cache because the cache key is the same across all of the pages.

<div class="clearfix"></div>

This mechanism has been working well from a performance perspective for a long
time. However, the time a website takes to respond to HTTP requests can reveal
that the browser has accessed the same resource in the past, which opens the
browser to security and privacy attacks, like the following:

> - **Detection if a user has visited a specific site**: An adversary can detect a

Reviewer: ```suggestion
- **Detect if a user has visited a specific site**: An adversary can detect a
```

  user's browsing history by checking if the cache has a resource which might be
  specific to a particular site or cohort of sites.
- **[Cross-site search
  attack](https://portswigger.net/daily-swig/new-xs-leak-techniques-reveal-fresh-ways-to-expose-user-information)**:
  An adversary can detect if an arbitrary string is in the user's search results
  by checking whether a 'no search results' image used by a particular website

>   is in the browser's cache or not.

Reviewer: ```suggestion
  is in the browser's cache.
```

Reviewer: Omit needless words (meaning words that don't contribute meaning).

- **Cross-site tracking**: The cache can be used to store cookie-like
  identifiers as a cross-site tracking mechanism.

To mitigate these risks, Chrome will partition its HTTP cache starting in Chrome 86.

## How will cache partitioning affect Chrome's HTTP Cache?

With cache partitioning, cached resources will be keyed using a new "Network
Isolation Key" in addition to the resource URL. The Network Isolation Key is
composed of the top-level site and the current-frame site.

> Note: The "site" is recognized using "scheme://eTLD+1" so if requests are from

Reviewer: I found information on eTLD. What does the "+1" mean? Is it something that needs an explanation.

Author: This is explained in the article linked at the last sentence.

Reviewer: In which link? I can't find it.

Author: This is within a "Note" section and at the bottom of the box it's linked to https://web.dev/same-site-same-origin/

> different pages, but they have the same scheme and effective top-level domain+1

Reviewer: Same question. What's "+1"?

Author: Same as above.

they will use the same cache partition. To learn more about this, read
[Understanding "same-site" and
"same-origin"](https://web.dev/same-site-same-origin/).

> Let's revisit the previous example to see how cache partitioning works in

Reviewer: ```suggestion
Look again at the previous example to see how cache partitioning works in
```

top-level -frame, the top-level site and current-frame site in the Network
Isolation Key are the same.)

<div class="clearfix"></div>

<figure class="attempt-left">
  <img src="/web/updates/images/2020/10/http-cache-partitioning-2.png">
  <figcaption>
    <b>Cache Key</b>: { <code>https://b.example</code>, <code>https://b.example</code>, <code>https://x.example/doge.png</code> }
  </figcaption>
</figure>

The same user visits a different page (`https://b.example`) which requests the
same image (`https://x.example/doge.png`). Though the same image was loaded in
the previous example, since the key doesn't match it will not be a cache hit.

The image is requested from the network and cached using `https://b.example` +
`https://b.example` + `https://x.example/doge.png` as the key.

<div class="clearfix"></div>

<figure class="attempt-left">
  <img src="/web/updates/images/2020/10/http-cache-partitioning-6.png">
  <figcaption>
    <b>Cache Key</b>: { <code>https://a.example</code>, <code>https://a.example</code>, <code>https://x.example/doge.png</code> }
  </figcaption>
</figure>

Now the user comes back to `https://a.example` but this time the image

> (`https://x.example/doge.png`) is embedded within an iframe. In this case, the

Reviewer: ```suggestion
(`https://x.example/doge.png`) is embedded in an iframe. In this case, the
```

key is `https://a.example` + `https://a.example` + `https://x.example/doge.png`
and a cache hit occurs. (Note that when the top-level site and the iframe are
the same site, the resource cached with the top-level frame can be used.

<div class="clearfix"></div>

<figure class="attempt-left">
  <img src="/web/updates/images/2020/10/http-cache-partitioning-4.png">
  <figcaption>
    <b>Cache Key</b>: { <code>https://a.example</code>, <code>https://c.example</code>, <code>https://x.example/doge.png</code> }
  </figcaption>
</figure>

The user is back at `https://a.example` but this time the image is hosted in an
iframe from `https://c.example`.

In this case, the image is downloaded from the network because there is no
resource in the cache that matches the key `https://a.example` +
`https://c.example` + `https://x.example/doge.png`.

<div class="clearfix"></div>

<figure class="attempt-left">
  <img src="/web/updates/images/2020/10/http-cache-partitioning-7.png">
  <figcaption>

>     <b>Cache Key</b>: { <code>https://a.example</code>, <code>https://c.example</code>, <code>https://x.example/doge.png</code> }

Reviewer: I think you should come up with a single way to represent keys and use it consistently.

Author: I don't understand your suggestion. I think it's already consistent?

Reviewer: In the previous paragraph you use + signs instead of showing it as a tuple.

Author: Again, please have a look at how this will look on the website: https://pr-9038-dot-web-central.appspot.com/web/updates/2020/10/http-cache-partitioning
I'm consistent on the tuples below the image while + on the paragraph to the right.

  </figcaption>
</figure>

What if the domain contains a subdomain or a port number? The user visits
`https://subdomain.a.example`, which embeds an iframe
(`https://c.example:8080`), which requests the image.

Because the key is created based on "scheme://eTLD+1", subdomains and port

> numbers are ignored, hence a cache hit occurs.

Reviewer: ```suggestion
numbers are ignored. Hence a cache hit occurs.
```

<div class="clearfix"></div>

<figure class="attempt-left">
  <img src="/web/updates/images/2020/10/http-cache-partitioning-5.png">
  <figcaption>
    <b>Cache Key</b>: { <code>https://a.example</code>, <code>https://c.example</code>, <code>https://x.example/doge.png</code> }
  </figcaption>
</figure>

What if the iframe is nested multiple times? The user visits
`https://a.example`, which embeds an iframe (`https://b.example`), which embeds
yet another iframe (`https://c.example`), which finally requests the image.

Because the key is taken from the top-frame (`https://a.example`) and the
immediate frame which loads the resource (`https://c.example`), a cache hit
occurs.

<div class="clearfix"></div>

> ## FAQs

Reviewer: This section can stay, but I offer a question as food for thought. If you answered these questions in the course of an article, why would anyone need to ask them?

Author: That's a fair point, but as this blog post has gone through approval process from PR, legal, marketing etc and took a loooong time, I'm hesitant to make significant changes. And I think the way it is now looks okay.

### As a web developer, are there any action I should take in response to this change?

This is not a breaking change, but it may impose performance considerations for

> some web services.

Reviewer: This, for example seems like information that should have been in your introduction.

Author: I read through the post once again, but I think it's ok from my POV.

For example, those that serve large volumes of highly cacheable resources across
many sites (such as fonts and popular scripts) may see an increase in their
traffic. Also, those who consume such services may have an increased reliance on
them.

(There's a proposal to enable shared libraries in a privacy-preserving way
called [Web Shared
Libraries](https://docs.google.com/document/d/1lQykm9HgzkPlaKXwpQ9vNc3m2Eq2hF4TY-Vup5wg4qg/edit#)
([presentation video](https://www.youtube.com/watch?v=cBY3ZcHifXw)), but it's
still under consideration.)

### What is the impact of this behavioral change?

The overall cache miss rate increases by about 3.6%, changes to the FCP (First

> Contentful Paint) are modest (~0.3%) and the overall fraction of bytes loaded

Reviewer: ```suggestion
Contentful Paint) are modest (~0.3%), and the overall fraction of bytes loaded
```

from the network increases by around 4%. You can learn more about the impact on
performance in [the HTTP cache partitioning
explainer](https://github.com/shivanigithub/http-cache-partitioning#impact-on-metrics).

### Is this standardized? Do other browsers behave differently?

"HTTP cache partitions" is [standardized in the fetch
spec](https://fetch.spec.whatwg.org/#http-cache-partitions) though browsers

> behave differently:

Reviewer: This strongly suggests to me that your article should be on web.dev. Did Kayce turn you down?

Author: I and @rowan-m agreed that this is Chrome specific change that doesn't require developer action and other browsers behave differently. @kaycebasques what is your thought?

Reviewer: For what its worth, its the fact that this is governed by a spec that makes me think web.dev. I'm not married to the idea.
